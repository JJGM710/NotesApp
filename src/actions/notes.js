import { db } from '../firebase/firebaseConfig';
import { types } from './../types/types';
import { loadNotes } from './../helpers/loadNotes';
import Swal from 'sweetalert2';
import { loadEntries } from './../helpers/loadEntries';
import { fileUpload } from './../helpers/fileUpload';

//accion asincrona que crea una nueva nota y luego hace el dispatch para que se active una nota, y se empiezan
// a cargar las nuevas entradas
export const startNewNote = () => {
	return async (dispatch, getState) => {
		const uid = getState().auth.uid;
		const newNote = {
			title: '',
			body: '',
			date: new Date().getTime(),
			url: null,
			update: null,
		};

		try {
			const doc = await db.collection(`${uid}/journal/notes`).add(newNote);
			dispatch(activeNote(doc.id, newNote));
			dispatch(startLoadingEntries(uid));
			dispatch(addNote(doc.id, newNote));
		} catch (error) {
			console.log(error);
		}

		// console.log(doc);
	};
};

//Carga las notas  y las coloca en el state del redux
export const startLoadingNotes = (uid) => {
	return async (dispatch, getState) => {
		const note = await loadNotes(uid);
		dispatch(setNotes(note));
	};
};

//carga el numero de entraddas que hay direectmaen desde la base de datos
//se puede optimizar para que sea lalmada de StartLoadingNotes
export const startLoadingEntries = (uid) => {
	return async (dispatch, getState) => {
		const entradas = await loadEntries(uid);
		dispatch(setEntry(entradas));
		console.log(entradas);
	};
};

//guarda la nota y hace el dispatch de refresh note para cambiar las propiedades de esa nota modificada
export const startSaveNote = (note) => {
	return async (dispatch, getState) => {
		const uid = getState().auth.uid;
		const { active: activeNote } = getState().notes;

		if (!note.url) {
			delete note.url;
		}

		const noteToFirestore = { ...note };
		delete noteToFirestore.id;
		//asignacion de propiedad para update y mandar a actualizar en la base de datos
		noteToFirestore.update = new Date().getTime();

		await db.doc(`${uid}/journal/notes/${note.id}`).update(noteToFirestore);
		//asignacion a la variable local luego de que se realizo el save en la base de datos
		activeNote.update = noteToFirestore.update;

		//se hace el refresh para que las  notas se actualicen con los nuevos cambios

		dispatch(refreshNote(note.id, noteToFirestore));
		Swal.fire('Saved', note.title, 'success');
	};
};

export const startUploading = (file) => {
	return async (dispatch, getState) => {
		const { active: activeNote } = getState().notes;

		Swal.fire({
			title: 'Uploading....',
			text: 'Please wait...',
			allowOutsideClick: false,
			onBeforeOpen: () => {
				Swal.showLoading();
			},
		});

		const fileUrl = await fileUpload(file);

		Swal.close();
		activeNote.url = fileUrl;
		dispatch(startSaveNote(activeNote));
	};
};

export const startDelete = (id) => {
	return async (dispatch, getState) => {
		const uid = getState().auth.uid;

		await db.doc(`${uid}/journal/notes/${id}`).delete();

		dispatch(deleNote(id));
		dispatch(startLoadingEntries(uid));
	};
};

//accciones sincronas
export const activeNote = (id, note) => {
	return {
		type: types.notesActive,
		payload: {
			id,
			...note,
		},
	};
};
export const setNotes = (notes) => {
	return {
		type: types.notesLoad,
		payload: notes,
	};
};

export const refreshNote = (id, note) => {
	return {
		type: types.notesUpdated,
		payload: {
			id,
			note: {
				id,
				...note,
			},
		},
	};
};

export const addNote = (id, note) => {
	return {
		type: types.notesAddNew,
		payload: {
			id,
			...note,
		},
	};
};

export const setEntry = (entradas) => {
	return {
		type: types.notesEntries,
		payload: entradas,
	};
};

export const deleNote = (id) => {
	return {
		type: types.notesDelete,
		payload: id,
	};
};

export const noteLogout = () => {
	return {
		type: types.notesLogoutCleaning,
	};
};
