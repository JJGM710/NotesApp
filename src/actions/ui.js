import { types } from './../types/types';
export const setError = (err) => {
	return {
		type: types.uiSetError,
		payload: err,
	};
};

export const removeError = () => {
	return {
		type: types.uiRemoveError,
	};
};

export const redirectError = () => {
	return {
		type: types.uiRedirect,
	};
};

export const removeRedirectError = () => {
	return {
		type: types.uiRemoveRedirect,
	};
};

export const startLoading = () => {
	return {
		type: types.uiStartLoading,
	};
};

export const finishLoading = () => {
	return {
		type: types.uiFinishLoading,
	};
};
