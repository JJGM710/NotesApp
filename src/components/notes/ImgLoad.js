import React, { useState } from 'react';
import Spinner from 'react-bootstrap/Spinner';

export const ImgLoad = ({ url }) => {
	const [loading, setLoading] = useState(false);
	const handleOnload = () => {
		setLoading(true);
	};

	return (
		<>
			{loading === false && (
				<div>
					<Spinner animation="grow" />
					<Spinner animation="grow" />
					<Spinner animation="grow" />
				</div>
			)}
			<div className="animate__animated animate__fadeIn ">
				<img
					src={url}
					alt="Imagen"
					className="notes__image "
					onLoad={handleOnload}
					style={{ display: loading ? 'block' : 'none' }}
				/>
			</div>
		</>
	);
};
