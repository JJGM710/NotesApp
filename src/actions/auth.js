import { types } from './../types/types';
import { firebase } from '../firebase/firebaseConfig';
import { googleAuthProvider } from '../firebase/firebaseConfig';
import {
	setError,
	finishLoading,
	startLoading,
	removeRedirectError,
} from './ui';
import Swal from 'sweetalert2';
import { noteLogout } from './notes';

export const startLogin = (email, password) => {
	return (dispatch) => {
		dispatch(startLoading());
		return firebase
			.auth()
			.signInWithEmailAndPassword(email, password)
			.then(({ user }) => {
				dispatch(login(user.uid, user.displayName));
				// dispatch(removeError());
				dispatch(finishLoading());
				Swal.fire('Logeed IN!', 'User Auth Correct', 'success');
			})
			.catch((e) => {
				// dispatch(setError(e.message));
				Swal.fire('Error', e.message, 'error');
				dispatch(finishLoading());
			});
	};
};

export const startRegister = (name, email, password) => {
	return (dispatch) => {
		dispatch(startLoading());
		return firebase
			.auth()
			.createUserWithEmailAndPassword(email, password)
			.then(async ({ user }) => {
				await user.updateProfile({ displayName: name });
				console.log(user.displayName);
				dispatch(login(user.uid, user.displayName));
				dispatch(finishLoading());
			})
			.catch((e) => {
				dispatch(finishLoading());
				Swal.fire('Error', e.message, 'error');
			});
	};
};

export const startGoogleLogin = () => {
	return (dispatch) => {
		firebase
			.auth()
			.signInWithPopup(googleAuthProvider)
			.then(({ user }) => {
				dispatch(login(user.uid, user.displayName));
			});
	};
};

export const login = (uid, displayName) => {
	return {
		type: types.login,
		payload: {
			uid,
			displayName,
		},
	};
};

export const startLogout = () => {
	return async (dispatch) => {
		dispatch(setError('Has cerrado sesion'));
		await firebase.auth().signOut();
		dispatch(removeRedirectError());
		dispatch(logout());
		dispatch(noteLogout());
	};
};

export const logout = () => ({
	type: types.logout,
});
