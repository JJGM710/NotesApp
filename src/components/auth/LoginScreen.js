import React, { useEffect, useState } from 'react';
import { Link } from 'react-router-dom';
import { useForm } from './../../hooks/useForm';
import { useDispatch, useSelector } from 'react-redux';
import { startLogin, startGoogleLogin } from './../../actions/auth';
import { removeError } from '../../actions/ui';

export const LoginScreen = React.memo(() => {
	const dispatch = useDispatch();
	const { loading } = useSelector((state) => state.ui);
	const { logout: estadolog } = useSelector((state) => state.auth);
	const [animate, setAnimate] = useState();

	useEffect(() => {
		setAnimate('animate__animated animate__fadeInUp animate__faster');
		return () => {
			if (estadolog === false) {
				dispatch(removeError());
			}
		};
	}, [dispatch, estadolog]);
	// console.log('Login Screen se llamo');

	const { formValues, handleInputChange } = useForm({
		email: '',
		password: '',
	});

	const { email, password } = formValues;

	const handleLogin = (e) => {
		e.preventDefault();
		console.log('login');
		dispatch(startLogin(email, password));
		// if (isloginValidationValid()) {
		// 	dispatch(startLogin(email, password));
		// }
	};

	const handleGoogleLogin = () => {
		// console.log('googlelogin');
		dispatch(startGoogleLogin());
	};

	// const isloginValidationValid = () => {
	// 	if (!validator.isEmail(email)) {
	// 		dispatch(setError('Invalid Email'));
	// 		return false;
	// 	}

	// return true;
	// };

	return (
		<>
			<h3 className="auth__title">Login</h3>
			<form onSubmit={handleLogin} className={animate}>
				<input
					type="text"
					placeholder="Email"
					name="email"
					autoComplete="off"
					className="auth__input"
					value={email}
					onChange={handleInputChange}
				/>
				<input
					type="password"
					placeholder="password"
					name="password"
					className="auth__input mt-3"
					autoComplete="current-password"
					value={password}
					onChange={handleInputChange}
				/>
				<button
					type="submit"
					className="btn btn-success btn-block mt-2"
					disabled={loading}
				>
					Login
				</button>

				<div className="auth__social-networks">
					<p>Login with Social Networks </p>

					<div className="google-btn" onClick={handleGoogleLogin}>
						<div className="google-icon-wrapper">
							<img
								className="google-icon"
								src="https://upload.wikimedia.org/wikipedia/commons/5/53/Google_%22G%22_Logo.svg"
								alt="google button"
							/>
						</div>
						<p className="btn-text">
							<b>Sign in with google</b>
						</p>
					</div>
				</div>
				<Link to="/auth/register" className="link">
					Create new account
				</Link>
			</form>
		</>
	);
});
