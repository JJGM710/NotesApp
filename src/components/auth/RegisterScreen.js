import React, { useEffect } from 'react';
import { Link } from 'react-router-dom';
import { useForm } from './../../hooks/useForm';
import validator from 'validator';
import { useDispatch, useSelector } from 'react-redux';
import { setError, removeError, removeRedirectError } from './../../actions/ui';
import { startRegister } from '../../actions/auth';

export const RegisterScreen = React.memo(() => {
	const dispatch = useDispatch();
	const { loading } = useSelector((state) => state.ui);

	const { formValues, handleInputChange } = useForm({
		name: 'Juan',
		email: 'juangiler7@outlook.com',
		password: '123456',
		password2: '123456',
	});

	const { name, email, password, password2 } = formValues;

	//activar papra eliminar el error del redirect cuando entre a registro
	// useEffect(() => {
	// 	dispatch(removeRedirectError());
	// }, [dispatch]);

	//manejo de mensajes con la pantalla de login screen

	useEffect(() => {
		dispatch(removeRedirectError());
		dispatch(removeError());
		return () => {
			dispatch(removeError());
		};
	}, [dispatch]);

	const handleRegister = (e) => {
		e.preventDefault();
		// console.log(name, email, password, password2);
		if (isFormValid()) {
			dispatch(startRegister(name, email, password));
		}
	};

	const isFormValid = () => {
		if (name.trim().length === 0) {
			dispatch(setError('Name is Required'));
			return false;
		} else if (!validator.isEmail(email)) {
			dispatch(setError('Enter valid Email'));
			return false;
		} else if (password !== password2 || password.length < 5) {
			dispatch(setError('Password Should be at 6 and match'));
			return false;
		}
		dispatch(removeError());
		return true;
	};
	/*
		{
			name: 'Juan'
			email: 'juangiler7@outlook.com'
			password: 123
			password2: 123
		}
		userForm

		const handleRegister 
		recibo el evento
		imprimir valores actualizados en click
	*/

	return (
		<>
			<h3 className="auth__title">Register</h3>

			<form
				onSubmit={handleRegister}
				className="animate__animated animate__fadeInUp animate__faster"
			>
				<input
					type="text"
					placeholder="Name"
					name="name"
					className="auth__input mt-5"
					value={name}
					onChange={handleInputChange}
				/>
				<input
					type="text"
					placeholder="Email"
					name="email"
					className="auth__input mt-5"
					value={email}
					onChange={handleInputChange}
				/>
				<input
					type="password"
					placeholder="Password"
					name="password"
					className="auth__input mt-5"
					autoComplete="current-password"
					value={password}
					onChange={handleInputChange}
				/>
				<input
					type="password"
					placeholder="Confirm Password"
					name="password2"
					className="auth__input mt-5"
					autoComplete="off"
					value={password2}
					onChange={handleInputChange}
				/>
				<button
					type="submit"
					className="btn btn-primary btn-block mt-1 mb-5"
					disabled={loading}
				>
					Register
				</button>

				<Link to="/auth/login" className="link">
					Already registered?
				</Link>
			</form>
		</>
	);
});
