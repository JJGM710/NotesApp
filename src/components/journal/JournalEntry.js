import React from 'react';
import moment from 'moment';
import { useDispatch } from 'react-redux';
import { activeNote } from '../../actions/notes';
export const JournalEntry = ({ id, date, title, body, url, update }) => {
	const noteDate = moment(date);
	console.log(url, update);
	const dispatch = useDispatch();
	const note = {
		title: title,
		body: body,
		date: date,
		url: url,
		update: update,
	};

	const handleEntreClick = () => {
		dispatch(activeNote(id, note));
	};

	const truncateString = (str, num) => {
		// If the length of str is less than or equal to num
		// just return str--don't truncate it.
		if (str.length <= num) {
			return str;
		}
		// Return str truncated with '...' concatenated to the end of str.
		return str.slice(0, num) + '...';
	};

	return (
		<div
			className="journal__entry animate__animated animate__fadeIn"
			onClick={handleEntreClick}
		>
			{url && (
				<div
					className="journal__entry-picture"
					style={{
						backgroundSize: 'cover',
						backgroundImage: `url(${url})`,
					}}
				></div>
			)}

			<div className="journal__entry-body">
				<p className="journal__entry-title">{truncateString(title, 30)}</p>
				<p className="journal__entry-content">{truncateString(body, 60)}</p>
			</div>

			<div className="journal__entry-date-box">
				<span>{noteDate.format('dddd')}</span>
				<h4>{noteDate.format('Do')}</h4>
			</div>
		</div>
	);
};
