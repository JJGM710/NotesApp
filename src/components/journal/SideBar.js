import React, { useState, useEffect } from 'react';
import { JournalEntries } from './JournalEntries';
import { useDispatch, useSelector } from 'react-redux';
import { startLogout } from './../../actions/auth';
import { startNewNote } from '../../actions/notes';

export const SideBar = () => {
	const dispatch = useDispatch();
	const { name } = useSelector((state) => state.auth);
	const { config } = useSelector((state) => state.notes);
	const [userbanner, setUserbanner] = useState();
	// console.log(config);
	const handleLogout = () => {
		dispatch(startLogout());
	};

	const handleAddNew = () => {
		dispatch(startNewNote());
	};

	useEffect(() => {
		if (config > 0) {
			setUserbanner(config);
		} else if (config === 0) {
			setUserbanner('No hay entradas');
		}
	}, [setUserbanner, config]);

	return (
		<aside className="journal__sidebar animate__animated animate__fadeIn animate__faster">
			<div className="journal__sidebar-navbar">
				<h3 className="mt-1">
					<i className="fas fa-user-astronaut"></i>
					<span>{`  ${name}  (${
						userbanner ? userbanner : 'Loading...'
					})`}</span>
				</h3>

				<button className="btn btn-danger" onClick={handleLogout}>
					Logout
				</button>
			</div>

			<div className="journal__new-entry" onClick={handleAddNew}>
				<i className="far fa-calendar-plus fa-3x"></i>
				<p className="mt-1"> New Entry</p>
			</div>

			<JournalEntries />
		</aside>
	);
};
