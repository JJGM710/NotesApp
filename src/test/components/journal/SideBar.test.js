import React from 'react';
import configureStore from 'redux-mock-store'; //ES6 modules
import thunk from 'redux-thunk';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import '@testing-library/jest-dom';
import { SideBar } from '../../../components/journal/SideBar';
import { startLogout } from './../../../actions/auth';
import { startNewNote } from './../../../actions/notes';

jest.mock('./../../../actions/auth', () => ({
	startLogout: jest.fn(),
}));

jest.mock('./../../../actions/notes', () => ({
	startNewNote: jest.fn(),
}));

const middlewares = [thunk];
const mockStore = configureStore(middlewares);
const initState = {
	auth: { uid: '1', name: 'JuanJose' },
	ui: { loading: false, msgError: null },
	notes: {
		notes: [],
		active: null,
		config: 2,
	},
};
let store = mockStore(initState);
store.dispatch = jest.fn();

describe('Pruebas en Sidebar', () => {
	const wrapper = mount(
		<Provider store={store}>
			<SideBar />
		</Provider>
	);
	test('debe mostrarse correctamente', () => {
		expect(wrapper).toMatchSnapshot();
	});

	test('debe llamar al startLogout', () => {
		wrapper.find('button').simulate('click');

		expect(startLogout).toHaveBeenCalled();
	});

	test('debe llamar el startNewNote', () => {
		wrapper.find('.journal__new-entry').simulate('click');

		expect(startNewNote).toHaveBeenCalled();
	});
});
