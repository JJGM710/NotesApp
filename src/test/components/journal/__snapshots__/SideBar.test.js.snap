// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en Sidebar debe mostrarse correctamente 1`] = `
<aside
  className="journal__sidebar animate__animated animate__fadeIn animate__faster"
>
  <div
    className="journal__sidebar-navbar"
  >
    <h3
      className="mt-1"
    >
      <i
        className="fas fa-user-astronaut"
      />
      <span>
          JuanJose  (2)
      </span>
    </h3>
    <button
      className="btn btn-danger"
      onClick={[Function]}
    >
      Logout
    </button>
  </div>
  <div
    className="journal__new-entry"
    onClick={[Function]}
  >
    <i
      className="far fa-calendar-plus fa-3x"
    />
    <p
      className="mt-1"
    >
       New Entry
    </p>
  </div>
  <div
    className="journal__entries "
  />
</aside>
`;
